src/rust/engine/graph/src/lib.rs

```diff
@@ -549,32 +619,27 @@ impl<N: Node> Graph<N> {
     C: NodeContext<Node = N>,
   {
     // Get or create the destination, and then insert the dep and return its state.
-    let dst_state = {
-      let mut inner = self.inner.lock().unwrap();
-      let dst_id = {
-        // TODO: doing cycle detection under the lock... unfortunate, but probably unavoidable
-        // without a much more complicated algorithm.
-        let potential_dst_id = inner.ensure_entry(EntryKey::Valid(dst_node.clone()));
-        if inner.detect_cycle(src_id, potential_dst_id) {
-          // Cyclic dependency: declare a dependency on a copy of the Node that is marked Cyclic.
-          inner.ensure_entry(EntryKey::Cyclic(dst_node))
-        } else {
-          // Valid dependency.
-          potential_dst_id
-        }
-      };
-
-      // Declare the dep, and return the state of the destination.
-      inner.pg.add_edge(src_id, dst_id, ());
-      inner
-        .entry_for_id_mut(dst_id)
-        .map(|entry| entry.state(context, dst_id))
-        .unwrap_or_else(|| future::err(N::Error::invalidated()).to_boxed().shared())
+    let mut inner = self.inner.lock().unwrap();
+    let dst_id = {
+      // TODO: doing cycle detection under the lock... unfortunate, but probably unavoidable
+      // without a much more complicated algorithm.
+      let potential_dst_id = inner.ensure_entry(EntryKey::Valid(dst_node.clone()));
```

Comment:
Would it be conceivable to only obtain a copy of the relevant subgraph under the lock (could use an r/w lock here too) with something like [filter_map](https://docs.rs/petgraph/*/petgraph/graph/struct.Graph.html#method.filter_map), do the cycle detection, then re-acquire the lock for mutation?

Reply:
So, the most basic form of that is hard, because acquiring a read lock and then "upgrading" it to a write lock is a rare, and fairly restrictive operation. The `RwLock` in the stdlib does not allow for it, and although the `RwLock` in the `parking_lot` crate does, you need to actually acquire explicitly as `upgradeable_read`, which is exclusive (just like acquiring for write).

I've also implemented a tactic involving adding a "mutation generation" to the graph that allowed us to cycle detect for a particular generation (not the same "generation" concept as on #6059) under the read lock, and then skip cycle detection if we acquired the write lock and the generation hadn't changed. And unfortunately that was _also_ not faster.

But this comment is sortof fud, because the most time I've ever seen cycle detection take in a profile is about 3%.